Update taint tracking docs (#821)
returntocorp/semgrep-docs: docs/writing-rules/data-flow/taint-mode.md

For example, in the code below, `some_safe_function` receives tainted data as input, and to be on the safe side, Semgrep assumes that it also returns tainted data as output. As a result, a finding is produced.

```javascript
var x = some_safe_function(tainted);
sink(x);
```

In some codebases, this conservative assumption can generate too many false positives. If that is the case, consider enumerating such functions as sanitizers:

```yaml
pattern-sanitizers:
- pattern: some_safe_function(...)
```

If this is too cumbersome, then set Semgrep to assume that functions are safe by default. This is done by setting `taint_assume_safe_functions: true` under the rule [`options`](/writing-rules/rule-syntax/#options) key. If you use this approach, you have to enumerate your taint propagators:

<iframe src="https://semgrep.dev/embed/editor?snippet=gBD0" border="0" frameBorder="0" width="100%" height="432"></iframe>

Similarly, by default, Semgrep assumes that indexing an array with a tainted index (that is, `array[tainted]`) is a tainted expression, even if the array itself is not tainted. Set `taint_assume_safe_indexes: true` to instead assume that these expressions are safe.

> If you want to have very fine-grained control over how taint is propagated, set `taint_only_propagate_through_assignments: true`, and then Semgrep only propagates taint through trivial assignments of the form `x = tainted`. In this case, you need to specify taint propagators or any other expression that can act as a taint propagator. For example, `x = taint_propagator(tainted)` or `x = tainted + "foo"` does not propagate any taint from `tainted` to `x` unless it is explicitly defined, whereas by default taint would propagate without explicit definition.

Author: ```suggestion
If you want to have very fine-grained control over how taint is propagated, set `taint_only_propagate_through_assignments: true`, and then Semgrep only propagates taint through trivial assignments of the form `x = tainted`. In this case, you need to specify taint propagators for any other expression that can act as a taint propagator. For example, with this setting `x = taint_propagator(tainted)` or `x = tainted + "foo"` do not propagate any taint from `tainted` to `x` unless it is explicitly specified, whereas by default taint would propagate implicitly.
```

# Taint tracking

Semgrep supports intra-procedural [taint tracking](https://en.wikipedia.org/wiki/Taint_checking). This data-flow analysis tracks the flow of untrusted (**tainted**) data throughout the body of a function or method. Tainted data originate from taint **sources**. If tainted data are not transformed or checked accordingly (**sanitized**), taint analysis reports a finding whenever tainted data reach a vulnerable function (what we call a **sink**). Tainted data flow from source to sink through **propagators** (assignments, function calls, etc).

The following video provides a quick overview of taint mode:
<iframe class="yt_embed" width="100%" height="432px" src="https://www.youtube.com/embed/6MxMhFPkZlU" frameborder="0" allowfullscreen></iframe>

Getting started

> ---------------

Reviewer: Optional note - should we rather change these headers to `## Getting started` (The same applies to Propagators and other second-level headers in this article - they are easier to search for.) @IagoAbal

Author: Fine with me, I just think `--------` it's a bit more visual. There should also exist editor extensions that let you navigate through subsections. But I don't spend that much time in the docs so feel free to change it.

Field sensitivity
-----------------

The taint engine provides basic field sensitivity support, it can:

- Track that `x.a.b` is tainted, but not `x` or `x.a`. If `x.a.b` is tainted, any extension of `x.a.b` (such as `x.a.b.c`) is considered tainted by default.

> - Track that `x.a` is tainted, but remember that `x.a.b` has been sanitized. Thus the engine knows that `x.a.b` is not tainted, but `x.a` or `x.a.c` are still tainted.

Author: I picked "records" instead.

:::

Sources
-------

A taint source is specified by a pattern, and _any_ subexpression that is matched by this pattern will be regarded as a source of taint.

:::note
Given the source specification below, and a piece of code such as `source(sink(x))`, the call `sink(x)` will be reported as a tainted sink.
The reason is that the pattern `source(...)` matches all of `source(sink(x))`, and that makes Semgrep label every subexpression in that piece of code as being a source. In particular, `x` is a source, and it is being passed into `sink`!

```yaml
pattern-sources:
- pattern: source(...)
```
:::

### Function arguments as sources

To specify that some argument of a function must be considered a taint source, simply write a pattern that matches that argument:

```javascript
pattern-sources:
  - patterns:
    - pattern-inside: function ($REQ, ...) {...}
    - focus-metavariable: $REQ
```

### Sources by side-effect

Consider the following hypothetical Python code where we call a `make_tainted` function that makes its argument tainted by side-effect:

```python
def make_tainted(a_set):
  a_set.add(tainted)

make_tainted(my_set)
sink(my_set)
```

> If we wrote the source specification below, the pattern formula would only match the occurrence of `x` that is the actual parameter of `make_tainted`, but it would not affect the occurrence of `x` that is passed to `sink`. They are the same variable, but they are different occurrences, and only the fist one is matched by the source specification. At the same time, Semgrep does not know that `make_tainted` is actually updating the variable `x` by side-effect. Thus, a taint rule using such specification would not produce any finding.

Reviewer: ```suggestion
Using the source specification below, the pattern formula only matches the occurrence of `x` that is the actual parameter of `make_tainted`, but it does not affect the occurrence of `x` that is passed to `sink`. Both are the same variable but in different occurrences. Only the first one is matched by the source specification. At the same time, Semgrep does not know that `make_tainted` is updating the variable `x` by side-effect. Thus, a taint rule using such specifications that are not producing any finding.
```

```yaml
pattern-sources:
- patterns:
  - pattern: make_tainted($X)
  - focus-metavariable: $X
```

> Until recently, the recommended workaround has been to write a source specification like the one below. This tells Semgrep that _every_ occurrence of `$X` after `make_tainted($X)` must be considered a source.

Reviewer: Maybe we should somehow mark this as sort of a **deprecated** workaround.

```suggestion
**Deprecated workaround**: The previously recommended workaround was to write a source specification such as the one specified below. This instructs Semgrep that **every** occurrence of `$X` after `make_tainted($X)` must be considered a source.
```

> Until recently, the recommended workaround has been to write a source specification like the one below. This tells Semgrep that _every_ occurrence of `$X` after `make_tainted($X)` must be considered a source.

Reviewer: Note (of course this is TW work, just so you comprehend the reasoning): We also should not use words such as "until recently" in docs - the document may not be updated for the next half a year and it will stay there...

```yaml
pattern-sources:
- patterns:
  - pattern: |
      make_tainted($X)
      ...
  - pattern: $X
```

> This has two main limitations. First, it will override any sanitization that may be performed on the code matched by `$X`.  In the example code below, the call `sink(x)` will be reported as tainted despite `x` has been re-assigned to a fresh set!

Reviewer: ```suggestion
This approach had two main limitations. First, it overrides any sanitization that can be performed on the code matched by `$X`.  In the example code below, the call `sink(x)` is reported as tainted despite `x` was re-assigned to a fresh set!
```

> This has two main limitations. First, it will override any sanitization that may be performed on the code matched by `$X`.  In the example code below, the call `sink(x)` will be reported as tainted despite `x` has been re-assigned to a fresh set!

Reviewer: Just a note - @IagoAbal we are including this comparison between the old and a new best practice so rule creators can update their rules, right?

Author: Yes. Also, some people may still prefer the old practice for whatever reason, but at least now they are more aware of the problems that it has.

Reviewer: Thanks! Just wanted to know the context.

```python
make_tainted(x)
x = set([])
sink(x) # false positive
```

> Also, we must note that [`...` has limitations](/writing-rules/pattern-syntax/#ellipses-and-statement-blocks) and, for example, in the code below we would not get any finding from Semgrep if we used that source specification:

Reviewer: ```suggestion
Note also that [`...` ellipses operator](/writing-rules/pattern-syntax/#ellipses-and-statement-blocks) has limitations. For example, in the code below Semgrep does not match any finding if such source specification is in use:
```

```python
if cond:
    make_tainted(x)
sink(x) # false negative
```

> The recommended solution nowadays is to set `by-side-effect: true`, being `by-side-effect` an option that is available for each individual source in a taint rule. When this option is enabled, _and_ the source specification matches a variable (or in general, an l-value) exactly, then Semgrep will assume that the variable (or l-value) becomes tainted by side-effect at the precise places where the source specification produces a match.

Reviewer: ```suggestion
It is now recommended to set `by-side-effect: true`, where `by-side-effect` is an option available for each individual source in a taint rule. When this option is enabled, and the source specification matches a variable (or in general, an l-value) exactly, then Semgrep assumes that the variable (or l-value) becomes tainted by side-effect at the precise places where the source specification produces a match.
```

> So, this is how we specify that `make_tainted` will update its argument making future occurrences tainted:

Reviewer: ```suggestion
So, this is how to specify that `make_tainted` updates its argument making future occurrences tainted:
```

<iframe src="https://semgrep.dev/embed/editor?snippet=eRvy" border="0" frameBorder="0" width="100%" height="432"></iframe>

> There are also [sanitizers by side effect](#sanitizers-by-side-effect). For a good example of use of both sources and sanitizers by side-effect, take a look at rule [`c.lang.security.use-after-free.use-after-free`](https://semgrep.dev/playground/r/c.lang.security.use-after-free.use-after-free).

Reviewer: ```suggestion
There are also [sanitizers by side effect](#sanitizers-by-side-effect). For a good example of the use of both sources and sanitizers by side-effect, take a look at rule [`c.lang.security.use-after-free.use-after-free`](https://semgrep.dev/playground/r/c.lang.security.use-after-free.use-after-free).
```

Propagators
-----------

> By default, tainted data automatically propagates through assignments, operators, and function calls (from inputs to output). However, there are other ways taint can propagate, which may require language- or library-specific knowledge that Semgrep does not have built in.

Reviewer: ```suggestion
By default, tainted data automatically propagates through assignments, operators, and function calls (from inputs to output). However, there are other ways in which taint can propagate, which can require language or library-specific knowledge that Semgrep does not have built-in.
```

Consider the following Python code where an unsafe `user_input` is stored into a `set` data structure. A random element from `set` is then passed into a `sink` function. This random element can be `user_input` itself, leading to an injection vulnerability!

```python
def test(s):
    x = user_input
    s = set([])
    s.add(x)
    #ruleid: test
    sink(s.pop())
```

> The following rule cannot find the above-described issue. The reason is that Semgrep does not know that executing `s.add(x)` makes `x` one of the elements in the set data structure `s`.

Reviewer: ```suggestion
The following rule cannot find the above-described issue. The reason is that Semgrep does not evaluate that executing `s.add(x)` makes `x` one of the elements in the set data structure `s`.
```

Author: Here `evaluate` sounds weird to me, I am proposing `is not aware`. Otherwise maybe something like "has no information about" is also OK.

```yaml
mode: taint
pattern-sources:
- pattern: user_input
pattern-sinks:
- pattern: sink(...)
```

The use of **taint propagators** enables Semgrep to propagate taint in this and other scenarios.
Taint propagators are specified under the `pattern-propagators` key:

```yaml
pattern-propagators:
- pattern: $S.add($E)
  from: $E
  to: $S
```

In the example above, Semgrep finds the pattern `$S.add($E)`, and it checks whether the code matched by `$E` is tainted. If it is tainted, Semgrep propagates that same taint to the code matched by `$S`. Thus, adding tainted data to a set marks the set itself as tainted.

<iframe src="https://semgrep.dev/embed/editor?snippet=dGRE" border="0" frameBorder="0" width="100%" height="432"></iframe>

> Another situation where taint propagatos may be useful is to specify in Java that, when we iterate a collection that is tainted, the individual elements must also be considered tainted:

Reviewer: ```suggestion
Another situation where taint propagators are useful is to specify in Java that, when we iterate a collection that is tainted, the individual elements must also be considered tainted:
```

> Another situation where taint propagatos may be useful is to specify in Java that, when we iterate a collection that is tainted, the individual elements must also be considered tainted:

Reviewer: @IagoAbal "we iterate a collection" - substitute "we" with something specific.

```yaml
pattern-propagators:
- pattern: $C.forEach(($X) -> ...)
  from: $C
  to: $X
```

In general, a taint propagator must specify:
1. A pattern containing **two** metavariables. These two metavariables specify where taint is propagated **from** and **to**.
2. The `to` and `from` metavariables. These metavariables should match a **variable** or an **expression**.
    - The `from` metavariable specifies the entry point of the taint.

>     - The `to` metavariable specifies where the tainted data is propagated to, typically an object or data structure.

Reviewer: @IagoAbal If this is the most important information for the reader - which it seems so - move it under the `Propagators-----------` header.

:::note

> When the `to` metavaraible matches a variable (or in general, an l-value), the taint will be propagated _by side-effect_.

Reviewer: ```suggestion
When the `to` metavariable matches a variable (or in general, an l-value), the taint is propagated _by side-effect_.
```

See [Sources by side effect](#sources-by-side-effect) for an explanation of what this means.
:::

> For example, pattern `$S.add($E)` includes two metavariables `$S` and `$E`. Given `from: $E` and `to: $S`, and with `$E` matching `x` and `$S` matching `s`, when `x` is tainted then `s` will become tainted (by side-effect) with the same taint as `x`.

Reviewer: ```suggestion
For example, pattern `$S.add($E)` includes two metavariables `$S` and `$E`. Given `from: $E` and `to: $S`, and with `$E` matching `x` and `$S` matching `s`, when `x` is tainted then `s` becomes tainted (by side-effect) with the same taint as `x`.
```

Sanitizers
-----------

> A sanitizer is specified by a pattern, and _any_ subexpression that is matched by this pattern will be regarded as sanitized.

Reviewer: ```suggestion
A sanitizer is specified by a pattern, and **any** subexpression that is matched by this pattern is regarded as sanitized.
```

:::note

> Given the sanitizer specification below, and a piece of code such as `sanitize(sink(source))`, the call `sink(source)` will _NOT_ be reported.

Reviewer: ```suggestion
Given the sanitizer specification below, and a piece of code such as `sanitize(sink(source))`, the call `sink(source)` is **not** reported.
```

The reason is that the pattern `sanitize(...)` matches all of `sanitize(sink(source))`, and that makes Semgrep label every subexpression in that piece of code as being sanitized. In particular, `source` is considered to be sanitized!

```yaml
pattern-sanitizers:
- pattern: sanitize(...)
```
:::

### Sanitizers by side-effect

Consider the following hypothetical Python code where we sanitize some tainted data by checking whether the data is safe to use, and raising an exception otherwise. It is then guaranteed that after `check_if_safe(x)`, the value of `x` must be a safe one.

```python
def check_if_safe(x):
  if not it_is_safe(x):
    raise Exception("unsafe")

x = tainted
check_if_safe(x)
sink(x)
```

> Until recently, the recommended workaround has been to write a sanitizer specification like the one below. This tells Semgrep that _every_ occurrence of `$X` after `check_if_safe($X)` must be considered safe.

Reviewer: ```suggestion
**Deprecated workaround**: The previously recommended workaround was to write a sanitizer specification such as the one below. This instructs Semgrep that **every** occurrence of `$X` after `check_if_safe($X)` must be considered safe.
```

pattern-sinks:
  - patterns:
    - pattern-inside: $ARRAY[$SINK]
    - focus-metavariable: $SINK
```

This way we tell Semgrep that we do not want arrays to be accessed with tainted indexes.

Minimizing false positives
--------------------------

> Since taint mode is intra-procedural, it does not know what other functions do. By default, Semgrep assumes that taint could always propagate through other functions.

Reviewer: ```suggestion
Since taint mode is intra-procedural, it does not recognize what other functions do. By default, Semgrep assumes that taint can always propagate through other functions.
```

Reviewer: Avoid modal verbs...

> For example, in the code below, `some_safe_function` receives tainted data as input and, to be on the safe side, Semgrep assumes that it will also return tainted data as output. As a result, a finding is produced.

Reviewer: ```suggestion
For example, in the code below, `some_safe_function` receives tainted data as input, and to be on the safe side, Semgrep assumes that it also returns tainted data as output. As a result, a finding is produced.
```

```javascript
var x = some_safe_function(tainted);
sink(x);
```

> In some codebases, this conservative assumption may produce too many false positives. If that is the case, you could first consider enumerating such functions as sanitizers:

Reviewer: ```suggestion
In some codebases, this conservative assumption can generate too many false positives. If that is the case, consider enumerating such functions as sanitizers:
```

```yaml
pattern-sanitizers:
- pattern: some_safe_function(...)
```

> If this is too cumbersome, then you can easily make Semgrep assume the opposite, that is, to assume that functions are safe by default. This is done by setting `taint_assume_safe_functions: true` under the rule [`options`](/writing-rules/rule-syntax/#options) key. If you use this approach, you instead have to enumerate your taint propagators:

Reviewer: ```suggestion
If this is too cumbersome, then set Semgrep to assume that functions are safe by default. This is done by setting `taint_assume_safe_functions: true` under the rule [`options`](/writing-rules/rule-syntax/#options) key. If you use this approach, you have to enumerate your taint propagators:
```

<iframe src="https://semgrep.dev/embed/editor?snippet=gBD0" border="0" frameBorder="0" width="100%" height="432"></iframe>

> Similarly, by default, Semgrep assumes that indexing an array with a tainted index (that is, `array[tainted]`) is a tainted expression, even if the array itself is not tainted. You can set `taint_assume_safe_indexes: true` to instead assume that these expressions are safe.

Reviewer: ```suggestion
Similarly, by default, Semgrep assumes that indexing an array with a tainted index (that is, `array[tainted]`) is a tainted expression, even if the array itself is not tainted. Set `taint_assume_safe_indexes: true` to instead assume that these expressions are safe.
```

> If you want to have very fine-grained control over how taint is propagated, you can set `taint_only_propagate_through_assignments: true`, and then Semgrep will only propagate taint via trivial assignemnts of the form `x = tainted`. You would need to specify taint propagators for anything else.

Reviewer: ```suggestion
If you want to have very fine-grained control over how taint is propagated, set `taint_only_propagate_through_assignments: true`. As a result, Semgrep only propagates taint through trivial assignments of the form `x = tainted`. In this case, you need to specify taint propagators for anything else.
```

> If you want to have very fine-grained control over how taint is propagated, you can set `taint_only_propagate_through_assignments: true`, and then Semgrep will only propagate taint via trivial assignemnts of the form `x = tainted`. You would need to specify taint propagators for anything else.

Reviewer: What do we mean by "anything else" here? @IagoAbal

Author: Any other expression that acts as a taint propagator. For example, `x = taint_propagator(tainted)` or `x = tainted + "foo"` won't propagate any taint from `tainted` to `x` unless you explicitly say so, whereas in the default setting this happens automatically.

Reviewer: I will push in a commit (as the diff here is outdated, no suggestion is possible), and substitute the text with the following (just a small change):

```
In this case, you need to specify taint propagators or any other expression that can act as a taint propagator.
```

Reviewer: @IagoAbal I will push in a commit (as the diff here is outdated, no suggestion is possible), and substitute the text with the following (just a small change):

```md
If you want to have very fine-grained control over how taint is propagated, set `taint_only_propagate_through_assignments: true`, and then Semgrep only propagates taint through trivial
assignments of the form `x = tainted`. In this case, you need to specify taint propagators or any other expression that can act as a taint propagator. For example, `x = taint_propagator(tainted)` or `x = tainted + "foo"` does not
propagate any taint from `tainted` to `x` unless it is explicitly defined, whereas in this happens automatically in the default taint analysis setting.
```

Reviewer: If this is misleading, feel free to rewrite this as you find best!

:::note

> Previously we recommended the use of so-called _not conflicting sanitizers_. This feature is no longer recommended and it should be considered deprecated.

Reviewer: ```suggestion
Previously, the use of so-called **not conflicting sanitizers** was recommended. This feature is now deprecated and no longer recommended.
```
